extensions/opc/src/fetchopc.cpp

```diff
@@ -0,0 +1,239 @@
+    if(idType_ != opc::OPCNodeIDType::Path) {
+      UA_NodeId myID;
+      myID.namespaceIndex = nameSpaceIdx_;
+      if(idType_ == opc::OPCNodeIDType::Int) {
+        myID.identifierType = UA_NODEIDTYPE_NUMERIC;
+        myID.identifier.numeric = std::stoi(nodeID_);
+      } else if (idType_ == opc::OPCNodeIDType::String) {
+        myID.identifierType = UA_NODEIDTYPE_STRING;
+        myID.identifier.string = UA_STRING_ALLOC(nodeID_.c_str());
+      }
+      opc::traverse(connection_, myID, f);
+    } else {
+      if(translatedNodeIDs_.empty()) {
+        auto sc = opc::translateBrowsePathsToNodeIdsRequest(connection_, nodeID_, translatedNodeIDs_, logger_);
+        if(sc != UA_STATUSCODE_GOOD) {
+          logger_->log_error("Failed to translate %s to node id, no flow files will be generated (%s)", nodeID_.c_str(), UA_StatusCode_name(sc));
+          yield();
+          return;
+        }
+      }
+      for(auto& nodeID: translatedNodeIDs_) {
+        opc::traverse(connection_, nodeID, f, nodeID_);
+      }
+    }
+    if(nodesFound_ == 0) {
+      logger_->log_warn("Connected to OPC server, the specified node was not found. Configuration might be incorrect! Yielding...");
+      yield();
+    } else if (variablesFound_ == 0) {
+      logger_->log_warn("Found no variables when traversing the specified node. No flowfiles are generated. Yielding...");
+      yield();
+    }
+
+  }
+
+  bool FetchOPCProcessor::nodeFoundCallBack(opc::ClientPtr& clientPtr, const UA_ReferenceDescription *ref, const std::string& path,
+      const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) {
+    nodesFound_++;
+    if(ref->nodeClass == UA_NODECLASS_VARIABLE)
+    {
+      try {
+        opc::nodeData nodedata = opc::getNodeData(clientPtr, ref);
+        OPCData2FlowFile(nodedata, context, session);
+        variablesFound_++;
+      } catch (const std::exception& exception) {
+        std::string browsename((char*)ref->browseName.name.data, ref->browseName.name.length);
+        logger_->log_warn("Caught Exception while trying to get data from node &s: %s", path + "/" + browsename,  exception.what());
+      }
+    }
+    return true;
+  }
+
+  void FetchOPCProcessor::OPCData2FlowFile(const opc::nodeData& opcnode, const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) {
+    std::shared_ptr<FlowFileRecord> flowFile = std::static_pointer_cast<FlowFileRecord>(session->create());
+    if (flowFile == nullptr) {
+      logger_->log_error("Failed to create flowfile!");
+      return;
+    }
+    for(const auto& attr: opcnode.attributes) {
+      flowFile->setAttribute(attr.first, attr.second);
+    }
+    if(opcnode.data.size() > 0) {
+      //if (extractValue_) {
+      try {
+        std::string value = opc::nodeValue2String(opcnode);
+        FetchOPCProcessor::WriteCallback callback(value);
```

Comment:
Can't we avoid this copy here? WriteCallback could store a const std::string& instead of copying it.

Reply:
Just changed it to move the value to the callback, adding it as an attribute was just a leftover anyway.